siege/siege.py

```diff
@@ -0,0 +1,194 @@
+# interleaved iterations. The two binaries are supplied on the commmand-line:
+#
+# Usage:
+#    siege.py clean_envoy_binary experimental_envoy_binary output_dir
+#
+# The two binaries should generally be built with
+#     bazel build -c opt source/exe:envoy-static
+# and can then be copied to a safe place (e.g. /tmp) so that they can each
+# be based on a single git workspace that you flip between branches.
+#
+# output_dir is where the logs and raw CSV files get written, but the
+# the summarized performance data is printed as a table to stdout.
+
+import json
+import os
+import requests
+import subprocess
+import sys
+import time
+import urllib.request
+
+# TODO(jmarantz): make the configuration pluggable, which means grepping
+# for these ports rather than hardcoding them. It's annoying to grep in
+# yaml files because there isn't per-line context. Another possibility
+# is to scrape them from Envoy log output.
+proxy_port = 10001
+upstream_port = 10000
+admin_port = 8081
+
+upstream_url = "http://127.0.0.1:%d" % upstream_port
+proxy_url = "http://127.0.0.1:%d" % proxy_port
+
+# We use the admin port both for scraping memory and for quitting at
+# the end of each run.
+admin = "http://127.0.0.1:%d" % admin_port
+
+def main(argv):
+  if len(argv) != 4:
+    print("Usage: %s clean_envoy_binary experimental_envoy_binary output_dir"
+          % argv[0])
+    sys.exit(1)
+
+  # Capture arguments as locals for readability.
+  clean_envoy = argv[1]
+  experimental_envoy = argv[2]
+  outdir = argv[3]
+  aggregate_csv= outdir + "/aggregate.csv"
+
+  # Connect to the script directory so local references to the yaml configs and
+  # lorem_ipsum.txt work.
+  siege_dir = os.path.abspath(os.path.dirname(argv[0]))
+  os.chdir(siege_dir)
+
+  # Derive some temp filenames from $outdir.
+  log = outdir + "/siege.log"
+  clean_perf_csv = outdir + "/clean.perf.csv"
+  clean_mem_csv = outdir + "/clean.mem.csv"
+  clean_csv_files = [clean_perf_csv, clean_mem_csv]
+  experimental_perf_csv = outdir + "/experimental.perf.csv"
+  experimental_mem_csv = outdir + "/experimental.mem.csv"
+  experimental_csv_files = [experimental_perf_csv, experimental_mem_csv]
+  aggregate_csv = outdir + "/aggregate.csv"
+  csv_files = clean_csv_files + experimental_csv_files + [aggregate_csv]
+  envoy_conf = siege_dir + "/front-proxy.yaml"
+
+  # Siege sometimes hangs when using time-based exit criteria.  See
+  # https://github.com/JoeDog/siege/issues/66.  In the meantime, use
+  # repetitions, which don't have that issue. If we were to use time
+  # we'd specify something like --time=5s.
+  #
+  # TODO(jmarantz): it would be nice to make the number of reps configurable,
+  # e.g. with a command-line option.
+  reps = 2000
+  siege_args = ["--reps=%d" % reps, "--rc=" + siege_dir + "/siege.conf", 
+                proxy_url]
+
+  # Clean up any old log files, so the CSVs just show the results from
+  # the runs we are about to do.
+  for fname in csv_files + [log]:
```

Comment:
Do we really need this step if doing `"w"` overwriting below?

Reply:
Not needed for the memory csv files because we are writing those. But if we don't clear the perf csvs then they'll be aggregated from previous runs by siege. I'll clarify in the comment.